Pub edit default form

allisonking opened the pull request:
Also see https://github.com/pubpub/platform/pull/821 for more in depth comments!

## Issue(s) Resolved

Part of https://github.com/pubpub/platform/issues/786

## High-level Explanation of PR
This mostly merges code paths between Pub editing and Pub editing via a form. 

| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/2b4470ba-7c43-450e-adcf-b6604efc3f39) | ![image](https://github.com/user-attachments/assets/87f07813-c445-4c49-bb80-68a1d69d9394)  |

There's still some repeated logic between `PubEditor` and the form fill page. So I think even more merging could be done, though I'm not quite sure the best way to go about it yet. If I get inspired in the next few days, I might try it, but also I don't know if I want to do it while @tefkah is changing all the `getPub` queries since that could get messy!

## Test Plan

Currently, because we don't have default forms in the db yet, you'll have to add a default form manually. I did this via Prisma, and set the seeded "Review" form in croccroc to `isDefault=true` 

* The new pub edit page at `/c/[communitySlug]/pubs/[pubid]/edit` should work
* The old form fill page should still work

Creating a pub is currently not working because there's no button in the modal anymore...

## Screenshots (if applicable)
![image](https://github.com/user-attachments/assets/5112d562-ddb3-49de-9c0c-e35d0b224590)


## Notes
The save button isn't as nice as it used to be because it is now outside of the form, which means we can't really query the form state to disable/make it loading. We have this same problem on the form builder page...

allisonking commented on core/app/components/StageSelect/StageSelectServer.tsx:
```diff
@@ -8,28 +8,12 @@ type Props = {
-	value?: StagesId;
 };
 
-export async function StageSelectServer({ communityId, fieldLabel, fieldName, value }: Props) {
```
I don't think passing the stage's value is actually needed here—the form takes care of that from what I can tell

allisonking commented on core/app/components/StageSelect/StageSelectClient.tsx:
```diff
@@ -24,16 +23,16 @@ export function StageSelectClient(props: Props) {
 						{...field}
-						value={props.stage?.id}
 						onValueChange={(value) => {
 							field.onChange(value);
 						}}
+						defaultValue={field.value}
```
I believe we should use the value from the form rather than passing a value in

tefkah replied:
makes sense to me!

allisonking commented on core/app/components/StageSelect/StageSelectClient.tsx:
```diff
@@ -24,16 +23,16 @@ export function StageSelectClient(props: Props) {
-							</SelectValue>
-						</SelectTrigger>
+						<FormControl>
+							<SelectTrigger>
+								<SelectValue placeholder="Select a stage" />
+							</SelectTrigger>
+						</FormControl>
```
this seems to work better, otherwise sometimes the selector is blank after making a new selection

allisonking commented on core/app/components/pubs/PubEditor/PubEditorClient.tsx:
```diff
@@ -1,348 +1,348 @@
 "use client";
```
this isn't a very useful diff—it would be more useful to diff this against `ExternalFormWrapper`! see diagram in PR description for more info

allisonking commented on core/app/c/(public)/[communitySlug]/public/forms/[formSlug]/fill/ExternalFormWrapper.tsx:
```diff
@@ -1,42 +1,16 @@
 "use client";
```
This isn't a useful diff— `ExternalFormWrapper` has totally changed. See diagram for info.

This and `PubEditorWrapper` both serve as client-side wrappers around `PubEditorClient`. They need a wrapper so that we can pass a function `onSuccess` as a prop, since both paths want to handle success states differently.

3mcd commented on core/lib/server/form.ts:
```diff
@@ -26,10 +26,15 @@ import { _getPubFields } from "./pubFields";
+	props: (
+		| { id: FormsId; slug?: never; pubTypeId?: never }
+		| { id?: never; slug: string; pubTypeId?: never }
+		| { id?: never; slug?: never; pubTypeId: PubTypesId }
```
Do you know the difference between this approach the following?

```ts
| { id: FormsId }
| { slug: string }
| { pubTypeId: PubTypesId }
```

Is it better intellisense? The former seems much more concise.

allisonking replied:
yup better intellisense, otherwise if you do a `props.slug` you get errors like

```
Property 'slug' does not exist on type '({ id: FormsId; } | { slug: string; } | { pubTypeId: PubTypesId; }) & { communityId: CommunitiesId; }'.
  Property 'slug' does not exist on type '{ id: FormsId; } & { communityId: CommunitiesId; }'
```

3mcd reviewed (approved):
Looks great! Took me a minute to figure out why the new default form wasn't showing up, but it ended up being a Next cache issue (obvious in retrospect). Nice work!

tefkah commented on core/app/c/[communitySlug]/pubs/[pubId]/edit/page.tsx:
```diff
@@ -0,0 +1,114 @@
+import { getCommunityBySlug } from "~/lib/db/queries";
+import { getPubsWithRelatedValuesAndChildren } from "~/lib/server";
+
+export async function generateMetadata({
```
I mentioned something similar in a PR from kalil, but could you add this `React.cache`d version of the function like this, and use it in both the body and the metadata?

This way, the result of the entire function call is stored in memory and deduped during the rendering of the page.

Eg we won't have to read from the filesystem cache and compute the cache keys twice. This is somewhat of a limitation to `autoCache`s current api. See here for a more thorough explanation: https://github.com/pubpub/platform/pull/813#discussion_r1869676750 

Not a blocker, you can ignore it for now if you just want to get it in, but I thought i'd mention it so we're all aware of this!

```suggestion
const getPubsWithRelatedValuesAndChildrenCached = cache(
	async (pubId: PubsId, communityId: CommunitiesId) => {
		return getPubsWithRelatedValuesAndChildren({
			pubId,
			communityId,
		});
	}
);

```

tefkah commented on core/app/c/[communitySlug]/pubs/[pubId]/edit/page.tsx:
```diff
@@ -0,0 +1,114 @@
+		redirect(`/c/${communitySlug}/unauthorized`);
+	}
+
+	const community = await getCommunityBySlug(communitySlug);
```
We should really get rid of this and rename the correct one, but could you use `findCommunityBySlug` instead? `getCommunityBySlug` still uses prisma and isn't cached!

tefkah commented on core/app/c/[communitySlug]/pubs/create/page.tsx:
```diff
@@ -0,0 +1,95 @@
+
+	const { user } = await getPageLoginData();
+
+	const community = await getCommunityBySlug(communitySlug);
```
Similar to on the edit page, could you use `findCommunityBySlug` instead?

tefkah commented on core/app/c/[communitySlug]/pubs/[pubId]/edit/page.tsx:
```diff
@@ -0,0 +1,114 @@
+}: {
+	params: { pubId: string; communitySlug: string };
+}): Promise<Metadata> {
+	const community = await getCommunityBySlug(communitySlug);
```
```suggestion
	const community = await findCommunityBySlug(communitySlug);
```

tefkah commented on core/app/c/[communitySlug]/pubs/create/page.tsx:
```diff
@@ -0,0 +1,95 @@
+}: {
+	params: { pubId: string; communitySlug: string };
+}): Promise<Metadata> {
+	const community = await getCommunityBySlug(communitySlug);
```
```suggestion
	const community = await findCommunityBySlug(communitySlug);
```

tefkah reviewed (approved):
Very nice work! This was quite a complex task, but you did it so quickly and so well, really great work!

None of the changes I'm requesting are real blockers, except for changing the uses of `getCommunityBySlug`. I'll approve the PR regardless so you could get it in today if you want, but that one would be nice to get in (and we should probably delete `getCommunityBySlug` to prevent these issues in the future!)


Also, in case you're interested, this PR provides a nice little speedboost on the `/pubs` and `/stages` page! (left old, right new)


https://github.com/user-attachments/assets/f4cf3666-3fca-4eb2-b8e3-55392934c4d1


https://github.com/user-attachments/assets/d73aeb09-cd39-4a05-97e8-6d86015bcfc9

With this, i will close #677 !

allisonking commented:
> Also, in case you're interested, this PR provides a nice little speedboost on the `/pubs` and `/stages` page! (left old, right new)

oh amazing, thanks for recording those!! ⚡ I think I've addressed all your comments, thanks for the thorough review!
